step_port_forward: add access to localhost if necessary

pdietl opened the pull request:
Closes #104

For Virtualbox versions >= 7.0, access to the host's network (via 10.0.2.2) is blocked by default unless the
`--nat-localhostreachableN=on` option is used.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/packer-plugin-virtualbox?pullRequest=122)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/packer-plugin-virtualbox?pullRequest=122)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/packer-plugin-virtualbox?pullRequest=122).</sub>

pdietl commented:
@nywilken could you take a look at this?

lbajolet-hashicorp reviewed (changes requested):
Hi @pdietl, 

Sorry about the time it took for us to take a look, that one flew under our radar, apologies for that.

I left a couple of comments on the code, I think it'd be clearer if the `addAccessToLocalhost` function returned an error instead of an action, then the caller can decide what to do with that error. There's also the semver creation from the constant, I believe this would be best to execute as part of the [`init`](https://golang.org/doc/effective_go.html#init) function, which is automatically invoked when the binary is executed. This would make it more robust as if for any reason the code doesn't work later on, we'll catch it proactively instead of when the step executes with the right version of virtualbox installed.

Aside from that, this looks good to me, please feel free to ping again when you've had time to address this feedback, and I'll take a second look then.

Thanks!

pdietl commented:
@lbajolet-hashicorp I believe I addressed all of your comments.

pdietl commented:
@lbajolet-hashicorp I can also add unit tests if you'd like.

lbajolet-hashicorp reviewed (commented):
Hi @pdietl,

Thank you for the reroll! In the current state the code looks good to me, I just left a last comment on the version being read, and there's some lint/compile errors in the step you added, when you address those I believe we'll be ready to merge this PR.

lbajolet-hashicorp commented:
> @lbajolet-hashicorp I can also add unit tests if you'd like.

If you have time for them, this would definitely be appreciated, thanks!

pdietl commented:
@lbajolet-hashicorp I added unit tests, formatted, and fixed compilation errors.
Here is some output of it working locally for me:
```
2024/02/28 17:41:17 packer-plugin-virtualbox plugin: 2024/02/28 17:41:17 VBoxManage --version output: 7.0.15r161716
2024/02/28 17:41:17 packer-plugin-virtualbox plugin: 2024/02/28 17:41:17 VirtualBox version: 7.0.15
2024/02/28 17:41:17 packer-plugin-virtualbox plugin: 2024/02/28 17:41:17 Executing VBoxManage: []string{"modifyvm", "packer-virtualbox-1709170875", "--nat-localhostreachable1", "on"}
2024/02/28 17:41:17 packer-plugin-virtualbox plugin: 2024/02/28 17:41:17 stdout:
2024/02/28 17:41:17 packer-plugin-virtualbox plugin: 2024/02/28 17:41:17 stderr:
2024/02/28 17:41:17 packer-plugin-virtualbox plugin: 2024/02/28 17:41:17 [TRACE] VirtualBox's version ("v7.0.15") is >= "v7.0.0", setting --nat-localhostreachable1 on
```

pdietl commented:
Also, I would appreciate it if you could release a new version after accepting this. :)

lbajolet-hashicorp commented on builder/virtualbox/common/step_port_forwarding.go:
```diff
@@ -34,6 +36,58 @@ type StepPortForwarding struct {
+	}
+}
+
+func getCanonicalSemver(version_string string) (string, error) {
```
Generally in Go it's preferred to use camelCase rather than snake_case for local/private entities.

lbajolet-hashicorp commented on builder/virtualbox/common/step_port_forwarding_test.go:
```diff
@@ -0,0 +1,100 @@
+const validSemver = "v7.0"
+const invalidSemver = "v7.0.0abcd"
+
+func TestValidSemver(t *testing.T) {
```
I'm not sure it's relevant to add this test since the code that does the conversion comes from another package, so as long as the dependency does the canonicalisation correctly, we shouldn't have any problem.

lbajolet-hashicorp commented on builder/virtualbox/common/step_port_forwarding.go:
```diff
@@ -34,6 +36,58 @@ type StepPortForwarding struct {
+func getCanonicalSemver(version_string string) (string, error) {
+	canonical_semver := semver.Canonical(version_string)
+	if canonical_semver == "" {
+		return "", fmt.Errorf("String '%s' is not a valid semver", version_string)
```
Instead of `'%s'` for quoted strings, you can use `%q` for this purpose

```suggestion
		return "", fmt.Errorf("String %q is not a valid semver", version_string)
```

lbajolet-hashicorp reviewed (commented):
Hi @pdietl,

We're getting there, I'm not sure we should unit-test the semver library, but aside from that this looks good to me.

I'll update the code to simplify this, and merge if it goes green directly after that.
Releasing the plugin after sounds good to me; I'll try to get it done today, or tomorrow at worst.

Thanks!
